pkg/services/k8s/apiserver/admission/schema_translate.go

```diff
@@ -5,88 +5,80 @@ import (
 	obj := a.GetObject()
+	ck := st.reg.ByName(obj.GetObjectKind().GroupVersionKind().Kind)
+	if ck == nil {
+		return nil
+	}
 
-	if obj.GetObjectKind().GroupVersionKind().Kind == "Dashboard" {
-		uobj := obj.(*unstructured.Unstructured)
-		spec, err := json.Marshal(uobj.Object["spec"])
-		if err != nil {
-			st.log.Error("failed to marshal spec", "err", err)
-			return nil
-		}
-		dk, err := dashboard.NewKind(cuectx.GrafanaThemaRuntime())
-		if err != nil {
-			st.log.Info("failed to create dashboard kind", "err", err)
-			return nil
-		}
-		sch, err := dk.Lineage().Schema(thema.LatestVersion(dk.Lineage()))
-		if err != nil {
-			st.log.Error("failed to get latest schema", "err", err)
-			return nil
-		}
+	uobj := obj.(*unstructured.Unstructured)
+	j, err := json.Marshal(uobj.Object)
+	if err != nil {
+		st.log.Error("failed to marshal spec to json", "err", err)
+		return nil
+	}
+
+	switch a.GetOperation() {
+	case admission.Create, admission.Update:
+		// TODO should we actually translate here? it won't guarantee resources are the current version in general, and we don't necessarily care. Should look at when webhook conversions are called...then be better, b/c those are terrible
 
-		switch a.GetOperation() {
-		case admission.Create, admission.Update:
-			cueVal, _ := vmux.NewJSONCodec("dashboard.json").Decode(cuectx.GrafanaCUEContext(), spec)
-			inst, err := sch.Validate(cueVal)
-			if err != nil {
-				st.log.Info("failed to validate dashboard", "err", err)
-				return nil
-			}
+		// JSON bytes in, translated JSON bytes out.
+		mux := vmux.NewByteMux(ck.Lineage().Latest(), vmux.NewJSONCodec(ck.MachineName()+".json"))
+		b, lac, err := mux(j)
 
-			// Translate doesn't return an error, so we just hope it doesn't panic.
-			ti, lacunas := inst.Translate(thema.LatestVersion(dk.Lineage()))
-			if ti == nil {
-				st.log.Info("failed to translate dashboard")
-				return nil
-			}
-			// not clear what to do if there are lacunas, so we just log them.
-			if lacunas != nil {
-				st.log.Info("failed to translate dashboard", "lacunas", lacunas)
-				return nil
-			}
+		if err != nil {
+			// For now, err will only ever be nil, because the validation handler already
+			// ensured the validity of the bytes. Once
+			// https://github.com/grafana/thema/issues/167 is in, errs will become possible
+			// directly from the translate call that happens in the vmuxer. (Currently,
+			// those error cases just panic :scream:)
+			return err
+		}
 
-			// how do I get from a thema instance back to a spec?
-			// for that matter, where do I put the translated spec?
-		case admission.Delete:
-			return nil
+		if lac != nil {
+			// TODO once we add a call that wraps in Go funcs to do lacuna handling, the only lacunas left here should be ones that are expected to make it to the user
+			st.log.Info("lacunas encountered during translation", lac)
 		}
+		uobj.Object = make(map[string]interface{})
+		json.Unmarshal(b, uobj.Object)
```

Comment:
What is this line doing? Are you checking for a panic? your `b` var isn't a pointer, so it's not unmarshaling properly (it also isn't used, but worth keeping an eye on if it does end up being important later)

Reply:
It's my first pass attempt - untested, since i'm still getting the other bits working -  at interpreting the k8s docs and some other code i found about how these mutating admission controllers are supposed to work. The thing we get from the `Admission` parameter to the func contains pointers to the underlying object, and it's expected that mutations are "returned" implicitly by updating the value of that pointer. Here's what `TaintNodesByCondition` [does](https://github.com/kubernetes/kubernetes/blob/c78204dc06d5b0bc02fc2f6bb7dbf98552180d26/plugin/pkg/admission/nodetaint/admission.go#L79-L95).

(i don't understand why we have an `*unstructured.Unstructured` from `Admission.GetObject()`, but that controller has an `*api.Node`...but one problem at a time?)